src/Components/Aspire.MySqlConnector/AspireMySqlConnectorExtensions.cs

```diff
@@ -0,0 +1,144 @@
+    }
+
+    private static void AddMySqlDataSource(IHostApplicationBuilder builder, string configurationSectionName,
+        Action<MySqlConnectorSettings>? configureSettings, string connectionName, object? serviceKey)
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+
+        MySqlConnectorSettings settings = new();
+        builder.Configuration.GetSection(configurationSectionName).Bind(settings);
+
+        if (builder.Configuration.GetConnectionString(connectionName) is string connectionString)
+        {
+            settings.ConnectionString = connectionString;
+        }
+
+        configureSettings?.Invoke(settings);
+
+        builder.RegisterMySqlServices(settings, configurationSectionName, connectionName, serviceKey);
+
+        // Same as SqlClient connection pooling is on by default and can be handled with connection string
+        // https://mysqlconnector.net/connection-options/#Pooling
+        if (settings.HealthChecks)
+        {
+            builder.TryAddHealthCheck(new HealthCheckRegistration(
+                serviceKey is null ? "MySql" : $"MySql_{connectionName}",
+                sp => new MySqlHealthCheck(new MySqlHealthCheckOptions()
+                {
+                    ConnectionString = serviceKey is null
+                        ? sp.GetRequiredService<MySqlDataSource>().ConnectionString
+                        : sp.GetRequiredKeyedService<MySqlDataSource>(serviceKey).ConnectionString
+                }),
+                failureStatus: default,
+                tags: default,
+                timeout: default));
+        }
+
+        if (settings.Tracing)
+        {
+            builder.Services.AddOpenTelemetry()
+                .WithTracing(tracerProviderBuilder =>
+                {
+                    tracerProviderBuilder.AddSource("MySqlConnector");
+                });
+        }
+
+        if (settings.Metrics)
+        {
+            builder.Services.AddOpenTelemetry()
+                .WithMetrics(meterProviderBuilder =>
+                {
+                    meterProviderBuilder.AddMeter("MySqlConnector");
+                });
+        }
+    }
+
+    private static void RegisterMySqlServices(this IHostApplicationBuilder builder, MySqlConnectorSettings settings, string configurationSectionName, string connectionName, object? serviceKey)
+    {
+        if (serviceKey is null)
+        {
+            // delay validating the ConnectionString until the DataSource is requested. This ensures an exception doesn't happen until a Logger is established.
+            builder.Services.AddMySqlDataSource(settings.ConnectionString ?? string.Empty, dataSourceBuilder =>
+            {
+                ValidateConnection();
+            });
+        }
+        else
+        {
+            // Currently MySqlConnector does not support Keyed DI Registration, so we implement it on our own.
```

Comment:
If I can implement https://github.com/mysql-net/MySqlConnector/issues/1391 then we can make the minimum requirement MySqlConnector >= 2.3.1 and simplify this code.

Reply:
Now using keyed services from MySqlConnector.DependencyInjection v2.3.1.